### ISSUE TITLE
S2-01 [Backend] Isolation Forest anomaly detectors — one model per country

### ISSUE THREAD

**Issue by NathanLuckock** [Nathan7108/HackathonPractice#15]
## Problem
Sentinel needs to detect when a country's behavior deviates from its historical baseline — sudden spike in conflict events, tone collapse, or anomalous event patterns. Without anomaly detection, the system only reports current state, not emerging threats. This is what makes Sentinel proactive rather than reactive.

## Goal
Train one Isolation Forest model per country on weekly GDELT aggregates. Save models to models/. Expose detect_anomaly(country_code, current_features) that returns anomaly score, flag, and severity.

### In scope
- train_anomaly_detector(country_code, gdelt_df) — trains one IsolationForest per country
- train_all_anomaly_detectors() — trains for all 8 countries
- detect_anomaly(country_code, current_features) -> {anomaly_score, is_anomaly, severity}
- Save models/anomaly_{CC}.pkl and models/scaler_{CC}.pkl for all 8 countries
- CLI runner: python -m backend.ml.anomaly trains all and tests Ukraine

### Out of scope
- Real-time continuous monitoring (hackathon: batch only)
- Integration with FastAPI (Sprint 3)
- Anomaly score injection into pipeline (done in FastAPI endpoint)

## Acceptance Criteria
- [ ] train_anomaly_detector("UA", gdelt_df) saves models/anomaly_UA.pkl and models/scaler_UA.pkl
- [ ] train_all_anomaly_detectors() produces 16 files (2 per country x 8 countries)
- [ ] detect_anomaly("UA", current_features) returns dict with anomaly_score (0-1), is_anomaly (bool), severity (LOW/MED/HIGH)
- [ ] anomaly_score normalized to 0-1 range
- [ ] severity: HIGH if score > 0.7, MED if > 0.4, LOW otherwise
- [ ] Handles missing model files gracefully (returns default LOW)
- [ ] Training completes in under 5 minutes total for all 8 countries

### 6 ANOMALY_FEATURES (weekly GDELT aggregates):
```python
ANOMALY_FEATURES = [
    'goldstein_mean',
    'goldstein_std',
    'goldstein_min',
    'mentions_total',
    'avg_tone',
    'event_count',
]
```

### Training logic (ML Guide Section 3.1):
```python
from sklearn.ensemble import IsolationForest
from sklearn.preprocessing import StandardScaler
import numpy as np
import pandas as pd
import joblib, os

def train_anomaly_detector(country_code: str, gdelt_df: pd.DataFrame):
    """
    Train one Isolation Forest per country.
    Learns country-specific 'normal' baseline from weekly aggregates.
    """
    df = gdelt_df.copy()
    df['date'] = pd.to_datetime(df['SQLDATE'].astype(str), format='%Y%m%d', errors='coerce')
    df['week'] = df['date'].dt.to_period('W')
    
    weekly = df.groupby('week').agg({
        'GoldsteinScale': ['mean', 'std', 'min'],
        'NumMentions': 'sum',
        'AvgTone': 'mean',
        'EventCode': 'count'
    }).reset_index()
    weekly.columns = ['week'] + ANOMALY_FEATURES
    weekly = weekly.fillna(0)
    
    X = weekly[ANOMALY_FEATURES].values
    scaler = StandardScaler()
    X_scaled = scaler.fit_transform(X)
    
    # contamination=0.05: ~5% of training weeks flagged as anomalies
    model = IsolationForest(n_estimators=200, contamination=0.05, random_state=42)
    model.fit(X_scaled)
    
    os.makedirs("models", exist_ok=True)
    joblib.dump(model, f"models/anomaly_{country_code}.pkl")
    joblib.dump(scaler, f"models/scaler_{country_code}.pkl")
    print(f"Trained anomaly detector for {country_code} on {len(weekly)} weeks")
    return model, scaler
```

### Detection logic:
```python
def detect_anomaly(country_code: str, current_features: dict) -> dict:
    """Run anomaly detection on current features."""
    try:
        model = joblib.load(f"models/anomaly_{country_code}.pkl")
        scaler = joblib.load(f"models/scaler_{country_code}.pkl")
    except FileNotFoundError:
        return {"anomaly_score": 0.0, "is_anomaly": False, "severity": "LOW"}
    
    X = np.array([[current_features.get(f, 0) for f in ANOMALY_FEATURES]])
    X_scaled = scaler.transform(X)
    raw_score = model.score_samples(X_scaled)[0]
    anomaly_score = max(0.0, min(1.0, (-raw_score - 0.3) / 0.7))
    is_anomaly = model.predict(X_scaled)[0] == -1
    
    return {
        "anomaly_score": round(float(anomaly_score), 3),
        "is_anomaly": bool(is_anomaly),
        "severity": "HIGH" if anomaly_score > 0.7 else "MED" if anomaly_score > 0.4 else "LOW"
    }
```

### Countries and data paths:
```python
COUNTRIES = ['UA', 'TW', 'IR', 'VE', 'PK', 'ET', 'RS', 'BR']
# GDELT CSVs at: data/gdelt/{CC}_events.csv
# Models saved to: models/anomaly_{CC}.pkl, models/scaler_{CC}.pkl
```

### GDELT CSV columns (as saved by fetch_gdelt.py):
The CSVs have headers: SQLDATE, Actor1CountryCode, Actor2CountryCode, EventCode, GoldsteinScale, NumMentions, AvgTone

### Edge cases:
- Country with very few events (e.g., Serbia, Brazil): may have < 10 weeks of data. Still train but warn.
- Country with no GDELT CSV: skip with warning, don't crash.
- GoldsteinScale may have NaN: fillna(0) before aggregation.
- models/ directory may not exist: create with os.makedirs.

### Hackathon tradeoff:
- contamination=0.05 is a reasonable default. In production you'd tune per country.
- Weekly aggregation loses daily granularity but is more stable with limited data.

## Manual Test Steps
1. python -m backend.ml.anomaly — trains all 8 countries, tests Ukraine
2. Verify models/ contains anomaly_UA.pkl, scaler_UA.pkl (and all other countries)
3. Verify detect_anomaly("UA", ...) returns a valid dict
4. Verify anomaly_score is between 0 and 1
5. Print number of training weeks per country

---

## Cursor Agent Instructions

Repo: Nathan7108/HackathonPractice
Working directory: C:\Users\natha\HackathonPractice
Branch: feature/s2-01-anomaly-detector

### STEP 0 — Read context
```
cat PROGRESS.md
cat backend/ml/data/fetch_gdelt.py
```
Understand GDELT CSV column names: SQLDATE, Actor1CountryCode, Actor2CountryCode, EventCode, GoldsteinScale, NumMentions, AvgTone

### STEP 1 — Create branch
```
git checkout main && git pull origin main
git checkout -b feature/s2-01-anomaly-detector
```

### STEP 2 — Implement backend/ml/anomaly.py
Replace placeholder with full implementation:
1. ANOMALY_FEATURES list (6 features)
2. COUNTRIES list (8 ISO2 codes)
3. train_anomaly_detector(country_code, gdelt_df) — weekly aggregation, StandardScaler, IsolationForest, save to models/
4. train_all_anomaly_detectors() — loop all 8 countries, load from data/gdelt/{CC}_events.csv
5. detect_anomaly(country_code, current_features) — load models, score, return dict
6. __main__ block that trains all and tests detect_anomaly on Ukraine

### STEP 3 — Validate
```
python -m backend.ml.anomaly
```
Verify: 16 model files created, Ukraine anomaly test works.

### STEP 4 — Commit and push
```
git add backend/ml/anomaly.py
git commit -m "feat: S2-01 Isolation Forest anomaly detectors for 8 countries"
git push -u origin feature/s2-01-anomaly-detector
```

### STEP 5 — Open PR
```
gh pr create --title "S2-01 [Backend] Isolation Forest anomaly detectors" --body "Closes #15" --base main
```

### Definition of Done
- [ ] train_all_anomaly_detectors() runs without error
- [ ] 16 files in models/ (anomaly_XX.pkl + scaler_XX.pkl x 8)
- [ ] detect_anomaly() returns valid dict with anomaly_score, is_anomaly, severity
- [ ] anomaly_score in [0, 1] range
- [ ] Handles missing countries gracefully
- [ ] Pushed to origin, PR created

### If You Get Stuck
- GDELT CSV path: data/gdelt/{CC}_events.csv (e.g., data/gdelt/UA_events.csv)
- GDELT columns: SQLDATE, GoldsteinScale, NumMentions, AvgTone, EventCode
- IsolationForest docs: sklearn.ensemble.IsolationForest
- If not enough data: reduce days_back when fetching GDELT, or accept fewer training weeks
